### PR TITLE
Code changes for SC parameters - zone and region

### PR DESCRIPTION
**Pull request by ambiknai**

### CODE REVIEW

File: pkg/ibmcsidriver/controller_helper.go [hunk 1]
```diff
@@ -271,6 +271,22 @@ func getVolumeParameters(logger *zap.Logger, req *csi.CreateVolumeRequest, confi
...
 
+	//if region is provided in storage class parameters, but zone is not given, error out
+	if len(strings.TrimSpace(volume.Az)) == 0 && len(strings.TrimSpace(volume.Region)) != 0 {
+		err = fmt.Errorf("zone parameter is not provided for the region - '%s':", strings.TrimSpace(volume.Region))
+		return volume, err
+	}
+
+	//if zone is provided but region is not provided, fetch region for specified zone
+	if len(strings.TrimSpace(volume.Az)) != 0 && len(strings.TrimSpace(volume.Region)) == 0 {
+		zones, err := pickTargetTopologyParams(req.GetAccessibilityRequirements())
```
Comment (sameshai) [inline pkg/ibmcsidriver/controller_helper.go:282]: Instead of writing 2 conditions can we combine something like this 

```
//If both zone and region not provided in storage class parameters then we pick from the Topology
	if len(strings.TrimSpace(volume.Region)) == 0  {
		zones, err := pickTargetTopologyParams(req.GetAccessibilityRequirements())
		if err != nil {
			err = fmt.Errorf("unable to fetch zone information: '%v'", err)
			logger.Error("getVolumeParameters", zap.NamedError("InvalidParameter", err))
			return volume, err
		}
		volume.Region = zones[utils.NodeRegionLabel]
       
               if len(strings.TrimSpace(volume.Az)) == 0  {
		volume.Az = zones[utils.NodeZoneLabel]
               }
	}
```

**Review by sameshai — APPROVED** [review]
/lgtm

**Comment by ambiknai** [comment]
```
git branch
  master
  parameters
  readme
* region
  revert
  scchanges

```


```
Successfully built 93747e3cd260
Successfully tagged gcr.io/k8s-staging-cloud-provider-ibm/ibm-vpc-block-csi-driver:amd64-dev
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# 
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# 
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# 
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# 
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# docker tag gcr.io/k8s-staging-cloud-provider-ibm/ibm-vpc-block-csi-driver:amd64-dev ambikanair/ibm-vpc-block-csi-driver:regchng
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# 
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# 
root@devmachine1:~/WORKSPACE/src/github.com/kubernetes-sigs/ibm-vpc-block-csi-driver# docker push ambikanair/ibm-vpc-block-csi-driver:regchng
The push refers to repository [docker.io/ambikanair/ibm-vpc-block-csi-driver]
3ff16a222296: Pushed 
014fc40a1eae: Pushed 
2cca63fa3287: Pushed 
cdc4d0cd3d15: Layer already exists 
43d05a2c9517: Layer already exists 
1902cf049561: Layer already exists 
9fe04650404a: Layer already exists 
regchng: digest: sha256:426f8cdfea0b2a4e8fb96e7207b4f5c3bbcc9ff888dafc4af67caa60e8fa1973 size: 1993
```

**Comment by ambiknai** [comment]
`TEST`

1. `Only zone is provided in SC`

```
oc describe sc testsczoneonly
Name:                  testsczoneonly
IsDefaultClass:        No
Annotations:           <none>
Provisioner:           vpc.block.csi.ibm.io
Parameters:            csi.storage.k8s.io/fstype=ext4,profile=10iops-tier,zone=us-south-2
AllowVolumeExpansion:  True
MountOptions:          <none>
ReclaimPolicy:         Retain
VolumeBindingMode:     Immediate
Events:                <none>
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc describe pvc testsczoneonly
Name:          testsczoneonly
Namespace:     default
StorageClass:  testsczoneonly
Status:        Bound
Volume:        pvc-10e76cc4-3dad-4ae9-877d-d8ae4bb3bb52
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: vpc.block.csi.ibm.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type    Reason                 Age               From                                                                                      Message
  ----    ------                 ----              ----                                                                                      -------
  Normal  Provisioning           24s               vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  External provisioner is provisioning volume for claim "default/testsczoneonly"
  Normal  ExternalProvisioning   8s (x3 over 24s)  persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "vpc.block.csi.ibm.io" or manually created by system administrator
  Normal  ProvisioningSucceeded  6s                vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  Successfully provisioned volume pvc-10e76cc4-3dad-4ae9-877d-d8ae4bb3bb52
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc describe pv pvc-10e76cc4-3dad-4ae9-877d-d8ae4bb3bb52
Name:              pvc-10e76cc4-3dad-4ae9-877d-d8ae4bb3bb52
Labels:            <none>
Annotations:       pv.kubernetes.io/provisioned-by: vpc.block.csi.ibm.io
Finalizers:        [kubernetes.io/pv-protection]
StorageClass:      testsczoneonly
Status:            Bound
Claim:             default/testsczoneonly
Reclaim Policy:    Retain
Access Modes:      RWO
VolumeMode:        Filesystem
Capacity:          10Gi
Node Affinity:     
  Required Terms:  
    Term 0:        failure-domain.beta.kubernetes.io/region in [us-south]
                   failure-domain.beta.kubernetes.io/zone in [us-south-2]
Message:           
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            vpc.block.csi.ibm.io
    FSType:            ext4
    VolumeHandle:      r006-b7867b85-851f-4008-aa94-bde428eab849
    ReadOnly:          false
    VolumeAttributes:      clusterID=c7scrg1d0f7m0kd3lc60
                           failure-domain.beta.kubernetes.io/region=us-south
                           failure-domain.beta.kubernetes.io/zone=us-south-2
                           iops=3000
                           storage.kubernetes.io/csiProvisionerIdentity=1644397023333-8081-vpc.block.csi.ibm.io
                           tags=
                           volumeCRN=crn:v1:bluemix:public:is:us-south-2:a/54e2afcb5f8eea01bfc3fde0ac755f1d::volume:r006-b7867b85-851f-4008-aa94-bde428eab849
                           volumeId=r006-b7867b85-851f-4008-aa94-bde428eab849
Events:
  Type    Reason               Age   From                            Message
  ----    ------               ----  ----                            -------
  Normal  VolumeMetaDataSaved  28s   ibm-vpc-block-csi-controller-0  Success
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % 


oc create -f ~/pod.yaml 
deployment.apps/testsczoneonly created
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc get pods | grep testsczoneonly                                       
testsczoneonly-55d4d85476-fmbhg   0/1     ContainerCreating   0          65s

oc get pods | grep testsczoneonly
testsczoneonly-55d4d85476-fmbhg   1/1     Running   0          111s

oc exec -it testsczoneonly-55d4d85476-fmbhg /bin/bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
root@testsczoneonly-55d4d85476-fmbhg:/# cd myvolumepath/
root@testsczoneonly-55d4d85476-fmbhg:/myvolumepath# 
root@testsczoneonly-55d4d85476-fmbhg:/myvolumepath# echo "hi" > new.txt
root@testsczoneonly-55d4d85476-fmbhg:/myvolumepath# exit
exit
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc get pods
NAME                              READY   STATUS    RESTARTS   AGE
hamster-79c9656654-2gxgj          1/1     Running   0          19h
hamster-79c9656654-8slrv          1/1     Running   0          19h
pod1                              1/1     Running   0          8d
pod2                              1/1     Running   0          8d
sconlyzone-56584589fb-rq8xr       1/1     Running   0          5d23h
sczoneandregion-6bc5f77b6-8h55x   1/1     Running   0          7d19h
testsczoneonly-55d4d85476-fmbhg   1/1     Running   0          15m
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc delete pod testsczoneonly-55d4d85476-fmbhg 
pod "testsczoneonly-55d4d85476-fmbhg" deleted
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc get pods
NAME                              READY   STATUS    RESTARTS   AGE
hamster-79c9656654-2gxgj          1/1     Running   0          19h
hamster-79c9656654-8slrv          1/1     Running   0          19h
pod1                              1/1     Running   0          8d
pod2                              1/1     Running   0          8d
sconlyzone-56584589fb-rq8xr       1/1     Running   0          5d23h
sczoneandregion-6bc5f77b6-8h55x   1/1     Running   0          7d19h
testsczoneonly-55d4d85476-2ml8q   1/1     Running   0          30s
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc exec -it testsczoneonly-55d4d85476-2ml8q /bin/bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
root@testsczoneonly-55d4d85476-2ml8q:/# ls /m
media/        mnt/          myvolumepath/ 
root@testsczoneonly-55d4d85476-2ml8q:/# ls /myvolumepath/
lost+found  new.txt
root@testsczoneonly-55d4d85476-2ml8q:/# 

```






2. `Only region is provided in SC`


```
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc create -f ~/sc.yaml 
storageclass.storage.k8s.io/testscregnonly created
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % vi ~/pvc.yaml 
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % 
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc create -f ~/pvc.yaml 
persistentvolumeclaim/testscregnonly created
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc get pvc | grep testscregnonly
testscregnonly        Pending                                                                        testscregnonly            20s
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc describe pvc testscregnonly
Name:          testscregnonly
Namespace:     default
StorageClass:  testscregnonly
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: vpc.block.csi.ibm.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type     Reason                Age               From                                                                                      Message
  ----     ------                ----              ----                                                                                      -------
  Warning  ProvisioningFailed    35s               vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  failed to provision volume with StorageClass "testscregnonly": rpc error: code = InvalidArgument desc = {RequestID: a45d70ca-4f75-4b5b-af1f-445f73f8c237 , Code: InvalidParameters, Description: Failed to extract parameters, BackendError: zone parameter is not provided for the region - 'us-south':, Action: Please provide valid parameters}
  Warning  ProvisioningFailed    34s               vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  failed to provision volume with StorageClass "testscregnonly": rpc error: code = InvalidArgument desc = {RequestID: 755ec6e7-2471-49d9-93c4-1f80e40c0ec8 , Code: InvalidParameters, Description: Failed to extract parameters, BackendError: zone parameter is not provided for the region - 'us-south':, Action: Please provide valid parameters}
  Warning  ProvisioningFailed    32s               vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  failed to provision volume with StorageClass "testscregnonly": rpc error: code = InvalidArgument desc = {RequestID: d06c3c06-bfa4-4ab3-a890-ece2ad337582 , Code: InvalidParameters, Description: Failed to extract parameters, BackendError: zone parameter is not provided for the region - 'us-south':, Action: Please provide valid parameters}
  Warning  ProvisioningFailed    28s               vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  failed to provision volume with StorageClass "testscregnonly": rpc error: code = InvalidArgument desc = {RequestID: 008530ec-473b-4f86-8da2-67f1497097f6 , Code: InvalidParameters, Description: Failed to extract parameters, BackendError: zone parameter is not provided for the region - 'us-south':, Action: Please provide valid parameters}
  Warning  ProvisioningFailed    20s               vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  failed to provision volume with StorageClass "testscregnonly": rpc error: code = InvalidArgument desc = {RequestID: c73176bb-f461-43aa-b6d4-cc99efb6edd2 , Code: InvalidParameters, Description: Failed to extract parameters, BackendError: zone parameter is not provided for the region - 'us-south':, Action: Please provide valid parameters}
  Normal   ExternalProvisioning  9s (x3 over 35s)  persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "vpc.block.csi.ibm.io" or manually created by system administrator
  Normal   Provisioning          4s (x6 over 35s)  vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  External provisioner is provisioning volume for claim "default/testscregnonly"
  Warning  ProvisioningFailed    4s                vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  failed to provision volume with StorageClass "testscregnonly": rpc error: code = InvalidArgument desc = {RequestID: 72c37928-8182-4c5a-90bc-f4a1597d415a , Code: InvalidParameters, Description: Failed to extract parameters, BackendError: zone parameter is not provided for the region - 'us-south':, Action: Please provide valid parameters}
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc describe sc testscregnonly
Name:                  testscregnonly
IsDefaultClass:        No
Annotations:           <none>
Provisioner:           vpc.block.csi.ibm.io
Parameters:            csi.storage.k8s.io/fstype=ext4,profile=10iops-tier,region=us-south
AllowVolumeExpansion:  True
MountOptions:          <none>
ReclaimPolicy:         Retain
VolumeBindingMode:     Immediate
Events:                <none>
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % 
```

3. `No zone and region provided in SC`

```
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc create -f ~/sc.yaml 
storageclass.storage.k8s.io/testscnorgnzone created
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc describe sc testscnorgnzone
Name:                  testscnorgnzone
IsDefaultClass:        No
Annotations:           <none>
Provisioner:           vpc.block.csi.ibm.io
Parameters:            csi.storage.k8s.io/fstype=ext4,profile=10iops-tier
AllowVolumeExpansion:  True
MountOptions:          <none>
ReclaimPolicy:         Retain
VolumeBindingMode:     Immediate
Events:                <none>
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % vi ~/pvc.yaml 
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc create -f ~/pvc.yaml
persistentvolumeclaim/testscnorgnzone created
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc get pvc| grep testscnorgnzone
testscnorgnzone       Bound     pvc-da4ba840-7f9c-4a07-b4c0-29802c3d3193   10Gi       RWO            testscnorgnzone           54s
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc describe pvc testscnorgnzone
Name:          testscnorgnzone
Namespace:     default
StorageClass:  testscnorgnzone
Status:        Bound
Volume:        pvc-da4ba840-7f9c-4a07-b4c0-29802c3d3193
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: vpc.block.csi.ibm.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type    Reason                 Age                From                                                                                      Message
  ----    ------                 ----               ----                                                                                      -------
  Normal  Provisioning           62s                vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  External provisioner is provisioning volume for claim "default/testscnorgnzone"
  Normal  ExternalProvisioning   54s (x2 over 62s)  persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "vpc.block.csi.ibm.io" or manually created by system administrator
  Normal  ProvisioningSucceeded  41s                vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_147f5adf-d7fe-403c-9e80-c855b930b739  Successfully provisioned volume pvc-da4ba840-7f9c-4a07-b4c0-29802c3d3193
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc describe pv pvc-da4ba840-7f9c-4a07-b4c0-29802c3d3193
Name:              pvc-da4ba840-7f9c-4a07-b4c0-29802c3d3193
Labels:            <none>
Annotations:       pv.kubernetes.io/provisioned-by: vpc.block.csi.ibm.io
Finalizers:        [kubernetes.io/pv-protection]
StorageClass:      testscnorgnzone
Status:            Bound
Claim:             default/testscnorgnzone
Reclaim Policy:    Retain
Access Modes:      RWO
VolumeMode:        Filesystem
Capacity:          10Gi
Node Affinity:     
  Required Terms:  
    Term 0:        failure-domain.beta.kubernetes.io/zone in [us-south-2]
                   failure-domain.beta.kubernetes.io/region in [us-south]
Message:           
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            vpc.block.csi.ibm.io
    FSType:            ext4
    VolumeHandle:      r006-ed76f2d2-f128-48a3-a35e-3f13c12a855b
    ReadOnly:          false
    VolumeAttributes:      clusterID=c7scrg1d0f7m0kd3lc60
                           failure-domain.beta.kubernetes.io/region=us-south
                           failure-domain.beta.kubernetes.io/zone=us-south-2
                           iops=3000
                           storage.kubernetes.io/csiProvisionerIdentity=1644397023333-8081-vpc.block.csi.ibm.io
                           tags=
                           volumeCRN=crn:v1:bluemix:public:is:us-south-2:a/54e2afcb5f8eea01bfc3fde0ac755f1d::volume:r006-ed76f2d2-f128-48a3-a35e-3f13c12a855b
                           volumeId=r006-ed76f2d2-f128-48a3-a35e-3f13c12a855b
Events:
  Type    Reason               Age   From                            Message
  ----    ------               ----  ----                            -------
  Normal  VolumeMetaDataSaved  47s   ibm-vpc-block-csi-controller-0  Success
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % 

oc create -f ~/pod.yaml
deployment.apps/testscnorgnzone created
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc get pods | grep testscnorgnzone
testscnorgnzone-85c5c4dd46-sgqgw   1/1     Running   0          109s
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc exec -it testscnorgnzone-85c5c4dd46-sgqgw  /bin/bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
root@testscnorgnzone-85c5c4dd46-sgqgw:/# cd /m
media/        mnt/          myvolumepath/ 
root@testscnorgnzone-85c5c4dd46-sgqgw:/# cd /m
media/        mnt/          myvolumepath/ 
root@testscnorgnzone-85c5c4dd46-sgqgw:/# cd /m
media/        mnt/          myvolumepath/ 
root@testscnorgnzone-85c5c4dd46-sgqgw:/# cd /myvolumepath/
root@testscnorgnzone-85c5c4dd46-sgqgw:/myvolumepath# 
root@testscnorgnzone-85c5c4dd46-sgqgw:/myvolumepath# echo "hi" > new.txt
root@testscnorgnzone-85c5c4dd46-sgqgw:/myvolumepath# exit
exit
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc delete pod testscnorgnzone-85c5c4dd46-sgqgw
pod "testscnorgnzone-85c5c4dd46-sgqgw" deleted
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc get pods | grep testscnorgnzone                     
testscnorgnzone-85c5c4dd46-5f5b9   1/1     Running   0          50s
ambikanair@Ambika-Nairs-MacBook-Pro ibm-vpc-block-csi-driver % oc exec -it testscnorgnzone-85c5c4dd46-5f5b9  /bin/bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
root@testscnorgnzone-85c5c4dd46-5f5b9:/# ls myvolumepath
lost+found  new.txt
root@testscnorgnzone-85c5c4dd46-5f5b9:/# 
```

**Review by prankulmahajan — COMMENTED** [review]
Can we also add the test case where both region and zone are not given?

**Comment by sameshai** [comment]
> Can we also add the test case in controller_helper_test where both region and zone are not given?

I think it is already there , @ambiknai can you confirm ?

**Comment by prankulmahajan** [comment]
> > Can we also add the test case in controller_helper_test where both region and zone are not given?
> 
> I think it is already there , @ambiknai can you confirm ?

The test case `“Zone and region not provided as parameter”,` is included in controller_test.go file, I was just confirming the same test case can be included in controller_helper_test.go as well. If we are covering max test cases in controller_test.go, then I am fine with that approach as well.

**Comment by ambiknai** [comment]
> > Can we also add the test case in controller_helper_test where both region and zone are not given?
> 
> I think it is already there , @ambiknai can you confirm ?
We dont have in "controller_helper_test" file. Can be included.  Will work on it along with Arashad's review comments

File: pkg/ibmcsidriver/controller_helper.go [hunk 2]
```diff
@@ -259,16 +259,25 @@ func getVolumeParameters(logger *zap.Logger, req *csi.CreateVolumeRequest, confi
...
-		volume.Az = zones[utils.NodeZoneLabel]
+		if len(strings.TrimSpace(volume.Az)) == 0 {
+			volume.Az = zones[utils.NodeZoneLabel]
+		}
+
+	}
+	//if region is provided in storage class parameters, but zone is not given, error out
+	if len(strings.TrimSpace(volume.Az)) == 0 && len(strings.TrimSpace(volume.Region)) != 0 {
+		err = fmt.Errorf("zone parameter is not provided for the region - '%s':", strings.TrimSpace(volume.Region))
+		return volume, err
```
Comment (arahamad) [inline pkg/ibmcsidriver/controller_helper.go:280]: if len(strings.TrimSpace(volume.Az)) == 0 && len(strings.TrimSpace(volume.Region)) != 0 {
		err = fmt.Errorf("zone parameter is not provided for the region - '%s':", strings.TrimSpace(volume.Region))
		return volume, err
	} 

       if  len(strings.TrimSpace(volume.Region)) == 0 {
		zones, err := pickTargetTopologyParams(req.GetAccessibilityRequirements())
		if err != nil {
			err = fmt.Errorf("unable to fetch zone information: '%v'", err)
			logger.Error("getVolumeParameters", zap.NamedError("InvalidParameter", err))
			return volume, err
		}
		volume.Region = zones[utils.NodeRegionLabel]
		if len(strings.TrimSpace(volume.Az)) == 0 {
			volume.Az = zones[utils.NodeZoneLabel]
		}
	}

**Review by arahamad — CHANGES_REQUESTED** [review]

File: pkg/ibmcsidriver/controller_helper_test.go [hunk 3]
```diff
@@ -301,6 +300,72 @@ func TestGetVolumeParameters(t *testing.T) {
...
+					Profile:       &provider.Profile{Name: "general-purpose"},
+					ResourceGroup: &provider.ResourceGroup{ID: "myresourcegroups"},
+				},
+				Region: "myregion",
+				Iops:   &noIops,
+				Az:     "myzone",
+			},
+			expectedStatus: true,
+			expectedError:  nil,
+		},
```
Comment (arahamad) [inline pkg/ibmcsidriver/controller_helper_test.go:368]: Can we add one more test case like `zone is given but region is empty`

**Review by arahamad — COMMENTED** [review]
run the golint tool as well

File: pkg/ibmcsidriver/controller_helper_test.go (hunk 3 above)
Comment (sameshai) [reply pkg/ibmcsidriver/controller_helper_test.go:368]: I think is it is already there 
https://github.com/kubernetes-sigs/ibm-vpc-block-csi-driver/pull/67/files#:~:text=name%3A%20%22Zone-,provided,-but%20region%20not

**Review by arahamad — APPROVED** [review]
/lgtm
